android/app/src/main/java/com/dismoi/scout/Floating/FloatingService.java

```diff
@@ -0,0 +1,193 @@
+package com.dismoi.scout.Floating;
+
+import android.app.Service;
+import android.content.Intent;
+import android.graphics.PixelFormat;
+import android.os.Binder;
+import android.os.Handler;
+import android.os.IBinder;
+import android.os.Looper;
+import android.view.Gravity;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.WindowManager;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.dismoi.scout.Floating.Layout.Bubble;
+import com.dismoi.scout.Floating.Layout.Message;
+import com.dismoi.scout.Floating.Layout.Trash;
+import com.dismoi.scout.Floating.Layout.Coordinator;
+import com.dismoi.scout.Floating.Layout.Base;
+
+public class FloatingService extends Service {
+  private final FloatingServiceBinder binder = new FloatingServiceBinder();
+  private final List<Bubble> bubbles = new ArrayList<>();
+  private final List<Message> messages = new ArrayList<>();
+  private Trash bubblesTrash;
+  private WindowManager windowManager;
+  private Coordinator layoutCoordinator;
+
+  @Override
+  public IBinder onBind(Intent intent) {
+    return binder;
+  }
+
+  @Override
+  public boolean onUnbind(Intent intent) {
+    for (Bubble bubble : bubbles) {
+      recycleBubble(bubble);
+    }
+    bubbles.clear();
+    return super.onUnbind(intent);
+  }
+
+  private void recycleBubble(final Bubble bubble) {
+    new Handler(Looper.getMainLooper()).post(new Runnable() {
+      @Override
+      public void run() {
+        getWindowManager().removeView(bubble);
+        for (Bubble cachedBubble : bubbles) {
+          if (cachedBubble == bubble) {
+            bubble.notifyBubbleRemoved();
+            bubbles.remove(cachedBubble);
+            break;
+          }
+        }
+      }
+    });
+  }
+
+  private void recycleMessage(final Message message) {
+    new Handler(Looper.getMainLooper()).post(new Runnable() {
+      @Override
+      public void run() {
+        getWindowManager().removeView(message);
+        for (Message cachedMessage : messages) {
+          if (cachedMessage == message) {
+            message.notifyBubbleRemoved();
+            messages.remove(cachedMessage);
+            break;
+          }
+        }
+      }
+    });
+  }
+
+  private WindowManager getWindowManager() {
+    if (windowManager == null) {
+      windowManager = (WindowManager)getSystemService(WINDOW_SERVICE);
+    }
+    return windowManager;
+  }
+
+  public void addBubble(Bubble bubble, int x, int y) {
```

Comment:
`addBubble` and `addDisMoiMessage` seems to be almost identical could it be factorized?

Reply:
I will factorize a little bit